src/cryptonote_core/loki_name_system.cpp

```diff
@@ -1134,73 +1188,127 @@ AND NOT EXISTS   (SELECT * FROM "mappings" WHERE "owner"."id" = "mappings"."back
+        }
+
+        save_settings(settings.top_height, settings.top_hash, static_cast<int>(db_version::v1_track_updates));
+      }
+    }
+  }
+
+  // ---------------------------------------------------------------------------
+  //
+  // Reupdate sql commands
+  //
+  // ---------------------------------------------------------------------------
+  // Old DB's can't pre-compile statements referencing new fields if they don't have it yet
+  if (DB_VERSION >= db_version::v1_track_updates)
   {
-    MERROR("Trying to apply non-existent transaction (no prior history of a db transaction beginning) to the LNS DB");
-    return;
+    char constexpr SAVE_MAPPING_STR_V2[] = R"(INSERT OR REPLACE INTO "mappings" ("type", "name_hash", "encrypted_value", "txid", "prev_txid", "register_height", "update_height", "owner_id", "backup_owner_id") VALUES (?,?,?,?,?,?,?,?,?))";
+    if (!sql_compile_statement(db, SAVE_MAPPING_STR_V2, loki::array_count(SAVE_MAPPING_STR_V2), &save_mapping_sql))
+      return false;
   }
 
-  char *sql_err = nullptr;
-  if (sqlite3_exec(lns_db.db, commit ? "END;" : "ROLLBACK;", NULL, NULL, &sql_err) != SQLITE_OK)
+  // ---------------------------------------------------------------------------
+  //
+  // Check settings
+  //
+  // ---------------------------------------------------------------------------
+  if (settings_record settings = get_settings())
   {
-    MERROR("Failed to " << (commit ? "end " : "rollback ") << " transaction to LNS DB, reason=" << (sql_err ? sql_err : "??"));
-    sqlite3_free(sql_err);
-    return;
+    if (!blockchain)
+    {
+      assert(nettype == cryptonote::FAKECHAIN);
+      return nettype == cryptonote::FAKECHAIN;
+    }
+
+    uint64_t lns_height   = 0;
+    crypto::hash lns_hash = blockchain->get_tail_id(lns_height);
+
+    // Try support out of date LNS databases by checking if the stored
+    // settings->[top_hash|top_height] match what we expect. If they match, we
+    // don't drop the DB but will load the missing blocks in a later step.
+
+    cryptonote::block lns_blk = {};
+    bool orphan               = false;
+    if (blockchain->get_block_by_hash(settings.top_hash, lns_blk, &orphan))
+    {
+      bool lns_height_matches = settings.top_height == cryptonote::get_block_height(lns_blk);
+      if (lns_height_matches && !orphan)
+      {
+        lns_height = settings.top_height;
+        lns_hash   = settings.top_hash;
+      }
+    }
+
+    if (settings.top_height == lns_height && settings.top_hash == lns_hash)
+    {
+      this->last_processed_height = settings.top_height;
+      assert(settings.version == static_cast<int>(DB_VERSION));
+    }
+    else
+    {
+      char constexpr DROP_TABLE_SQL[] = R"(DROP TABLE IF EXISTS "owner"; DROP TABLE IF EXISTS "settings"; DROP TABLE IF EXISTS "mappings")";
+      sqlite3_exec(db, DROP_TABLE_SQL, nullptr /*callback*/, nullptr /*callback context*/, nullptr);
```

Comment:
Does this multi-query execute work okay in sqlite3?  I know some databases only allow a single statement per query execution.

Reply:
Can confirm, works (tried it out).